File: src/cli/src/session.ts
import {
  AcpError,
  KernelDisconnected,
  type AcpClient,
  type ExecutionResult,
  type PromptResult,
  type SessionUpdateParams,
} from "@/acp/client.js";
import { AcpMethod, MustangMethod } from "@/acp/methods.js";
import type { CliSessionInfo } from "@/sessions/types.js";
import { randomUUID } from "node:crypto";
import { cwd } from "process";

type MustangSessionClient = Pick<
  AcpClient,
  "request" | "notify" | "promptRequest" | "activateSkillRequest" | "executeShellRequest" | "executePythonRequest" | "onUpdate"
>;

export type PermissionMode = "default" | "accept_edits" | "plan" | "auto" | "dont_ask" | "bypass";

type SessionModeState = {
  configOptions?: unknown;
  config_options?: unknown;
  modes?: unknown;
};

export interface CostUsageReport {
  sessionId: string;
  title?: string | null;
  cwd: string;
  createdAt?: string | null;
  updatedAt?: string | null;
  model?: string | null;
  kernelVersion: string;
  tokens: { input: number; output: number; cacheRead: number; cacheWrite: number; total: number };
  context: {
    totalTokens: number;
    contextWindow?: number | null;
    percent: number;
    sections: Array<{ id: string; label: string; tokens: number; percent: number }>;
  };
  history: {
    messages: number;
    turns: number;
    toolCalls: number;
    compactions: number;
    queuedTurns: number;
    inFlight: boolean;
    lastRunAt?: string | null;
    lastDurationMs?: number | null;
  };
  memory: { loaded: number; writableScopes: number };
  environment: { lspServers: string[]; mcpServers: string[] };
  costUsd?: number | null;
  costNote?: string | null;
}

export interface CommandEntry {
  name: string;
  description: string;
  usage: string;
  acpMethod?: string | null;
  acp_method?: string | null;
  subcommands?: string[];
  source?: string;
}

export interface ListCommandsResponse {
  commands: CommandEntry[];
}

export interface RuntimeStatusReport {
  status: Record<string, unknown>;
}

const RESUME_RETRY_ATTEMPTS = 24;
const RESUME_RETRY_DELAY_MS = 250;

export class MustangSession {
  constructor(
    private client: MustangSessionClient,
    public readonly sessionId: string,
    public summary?: CliSessionInfo,
  ) {}

  static async create(
    client: AcpClient,
    workingDir?: string,
  ): Promise<MustangSession> {
    const result = await client.request<{ sessionId: string }>(AcpMethod.sessionNew, {
      cwd: workingDir ?? cwd(),
      mcpServers: [],
    });
    return new MustangSession(client, result.sessionId);
  }

  static async load(
    client: AcpClient,
    id: string,
    workingDir?: string,
  ): Promise<MustangSession> {
    await client.request(AcpMethod.sessionLoad, {
      sessionId: id,
      cwd: workingDir ?? cwd(),
      mcpServers: [],
    });
    return new MustangSession(client, id);
  }

  static async getUsage(client: AcpClient, sessionId?: string): Promise<CostUsageReport> {
    const params = sessionId
      ? {
          sessionId,
          session_id: sessionId,
        }
      : {};
    return await client.request<CostUsageReport>(MustangMethod.sessionGetUsage, params);
  }

  async prompt(
    text: string,
    onUpdate: (update: SessionUpdateParams) => void,
    options: { mode?: PermissionMode } = {},
  ): Promise<PromptResult> {
    const unsub = this.client.onUpdate(onUpdate);
    const clientTurnId = randomUUID();
    try {
      let resumeState = await this.resumeWithRetry();
      await this.syncModeAfterResume(options.mode, resumeState);
      try {
        return await this.client.promptRequest(this.sessionId, text, { clientTurnId });
      } catch (error) {
        if (!(error instanceof KernelDisconnected)) throw error;
        resumeState = await this.resumeWithRetry();
        await this.syncModeAfterResume(options.mode, resumeState);
        return await this.client.promptRequest(this.sessionId, text, { clientTurnId });
      }
    } finally {
      unsub();
    }
  }

  async activateSkill(
    skill: string,
    args: string,
    onUpdate: (update: SessionUpdateParams) => void,
    options: { mode?: PermissionMode } = {},
  ): Promise<PromptResult> {
    const unsub = this.client.onUpdate(onUpdate);
    const clientTurnId = randomUUID();
    try {
      let resumeState = await this.resumeWithRetry();
      await this.syncModeAfterResume(options.mode, resumeState);
      try {
        return await this.client.activateSkillRequest(this.sessionId, skill, args, { clientTurnId });
      } catch (error) {
        if (!(error instanceof KernelDisconnected)) throw error;
        resumeState = await this.resumeWithRetry();
        await this.syncModeAfterResume(options.mode, resumeState);
        return await this.client.activateSkillRequest(this.sessionId, skill, args, { clientTurnId });
      }
    } finally {
      unsub();
    }
  }

  async listCommands(): Promise<CommandEntry[]> {
    const result = await this.client.request<ListCommandsResponse>(
      MustangMethod.commandsList,
      {},
    );
    return result.commands ?? [];
  }

  async executeShell(
    command: string,
    excludeFromContext: boolean,
    onUpdate: (update: SessionUpdateParams) => void,
  ): Promise<ExecutionResult> {
    const unsub = this.client.onUpdate(onUpdate);
    try {
      return await this.client.executeShellRequest(this.sessionId, command, excludeFromContext);
    } finally {
      unsub();
    }
  }

  async executePython(
    code: string,
    excludeFromContext: boolean,
    onUpdate: (update: SessionUpdateParams) => void,
  ): Promise<ExecutionResult> {
    const unsub = this.client.onUpdate(onUpdate);
    try {
      return await this.client.executePythonRequest(this.sessionId, code, excludeFromContext);
    } finally {
      unsub();
    }
  }

  cancel(): void {
    this.client.notify(AcpMethod.sessionCancel, { sessionId: this.sessionId });
  }

  cancelExecution(kind: "shell" | "python" | "any" = "any"): void {
    this.client.notify(MustangMethod.sessionCancelExecution, { sessionId: this.sessionId, kind });
  }

  async setMode(mode: PermissionMode): Promise<void> {
    await this.client.request(AcpMethod.sessionSetMode, {
      sessionId: this.sessionId,
      modeId: mode,
    });
  }

  async getUsage(): Promise<CostUsageReport> {
    // Send both spellings for compatibility with already-running dev kernels
    // that may have loaded the new method before the ACP camelCase base model.
    return await MustangSession.getUsage(this.client as AcpClient, this.sessionId);
  }

  async runtimeStatus(): Promise<RuntimeStatusReport> {
    return await this.client.request<RuntimeStatusReport>(MustangMethod.runtimeStatus, {});
  }

  async runtimeRestart(reason = "user requested runtime restart"): Promise<RuntimeStatusReport> {
    return await this.client.request<RuntimeStatusReport>(MustangMethod.runtimeRestart, { reason });
  }

  private async resumeWithRetry(): Promise<SessionModeState> {
    let lastError: unknown;
    for (let attempt = 1; attempt <= RESUME_RETRY_ATTEMPTS; attempt++) {
      try {
        return await this.resume();
      } catch (error) {
        lastError = error;
        if (!isTransientResumeError(error) || attempt === RESUME_RETRY_ATTEMPTS) break;
        await sleep(RESUME_RETRY_DELAY_MS);
      }
    }
    throw lastError;
  }

  private async resume(): Promise<SessionModeState> {
    return await this.client.request<SessionModeState>(AcpMethod.sessionResume, {
      sessionId: this.sessionId,
      cwd: cwd(),
    });
  }

  private async syncModeAfterResume(
    desiredMode: PermissionMode | undefined,
    resumeState: SessionModeState,
  ): Promise<void> {
    if (!desiredMode) return;
    if (extractPermissionMode(resumeState) === desiredMode) return;
    await this.setMode(desiredMode);
  }
}

function isTransientResumeError(error: unknown): boolean {
  return error instanceof AcpError && error.code === -32603 && error.message.includes("Internal error");
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function extractPermissionMode(value: unknown): PermissionMode | undefined {
  const item = value as Record<string, unknown> | undefined;
  const modes = item?.modes as { currentModeId?: unknown; current_mode_id?: unknown } | undefined;
  const fromModes = parsePermissionMode(modes?.currentModeId ?? modes?.current_mode_id);
  if (fromModes) return fromModes;

  const configOptions = Array.isArray(item?.configOptions)
    ? item.configOptions
    : Array.isArray(item?.config_options)
      ? item.config_options
      : undefined;
  const modeConfig = configOptions?.find(option => {
    const record = option as { configId?: unknown; config_id?: unknown } | undefined;
    return record?.configId === "mode" || record?.config_id === "mode";
  }) as { currentValue?: unknown; current_value?: unknown } | undefined;
  const fromConfig = parsePermissionMode(modeConfig?.currentValue ?? modeConfig?.current_value);
  if (fromConfig) return fromConfig;

  return item && "raw" in item ? extractPermissionMode(item.raw) : undefined;
}

function parsePermissionMode(value: unknown): PermissionMode | undefined {
  const modes: PermissionMode[] = ["default", "accept_edits", "plan", "auto", "dont_ask", "bypass"];
  return typeof value === "string" && modes.includes(value as PermissionMode)
    ? value as PermissionMode
    : undefined;
}

File: src/kernel/kernel/protocol/acp/schemas/session.py
"""ACP session method wire-format schemas."""

from __future__ import annotations

from typing import Any
from typing import Literal

from kernel.protocol.acp.schemas.base import AcpModel
from kernel.protocol.acp.schemas.content import AcpContentBlock
from kernel.protocol.acp.schemas.enums import AcpStopReason


class AcpMcpServer(AcpModel):
    name: str
    command: str | None = None
    args: list[str] = []
    env: list[dict[str, str]] = []
    type: str | None = None
    url: str | None = None
    headers: list[dict[str, str]] = []


class AcpSessionInfo(AcpModel):
    session_id: str
    cwd: str
    updated_at: str
    title: str | None = None
    archived_at: str | None = None
    title_source: Literal["auto", "user"] | None = None
    meta: dict[str, Any] | None = None


# session/new


class NewSessionRequest(AcpModel):
    cwd: str
    mcp_servers: list[AcpMcpServer] = []
    meta: dict[str, Any] | None = None


class NewSessionResponse(AcpModel):
    session_id: str
    config_options: list[dict] | None = None
    modes: dict | None = None
    meta: dict[str, Any] | None = None


# session/load


class LoadSessionRequest(AcpModel):
    session_id: str
    cwd: str
    mcp_servers: list[AcpMcpServer] = []
    meta: dict[str, Any] | None = None


class LoadSessionResponse(AcpModel):
    config_options: list[dict] | None = None
    modes: dict | None = None
    meta: dict[str, Any] | None = None


# session/resume


class ResumeSessionRequest(AcpModel):
    session_id: str
    cwd: str | None = None
    meta: dict[str, Any] | None = None


class ResumeSessionResponse(AcpModel):
    config_options: list[dict] | None = None
    modes: dict | None = None
    meta: dict[str, Any] | None = None


# session/close


class CloseSessionRequest(AcpModel):
    session_id: str
    meta: dict[str, Any] | None = None


class CloseSessionResponse(AcpModel):
    meta: dict[str, Any] | None = None


# session/list


class ListSessionsRequest(AcpModel):
    cursor: str | None = None
    cwd: str | None = None
    include_archived: bool = False
    archived_only: bool = False
    meta: dict[str, Any] | None = None


class ListSessionsResponse(AcpModel):
    sessions: list[AcpSessionInfo]
    next_cursor: str | None = None
    meta: dict[str, Any] | None = None


# session/prompt


class PromptRequest(AcpModel):
    session_id: str
    prompt: list[AcpContentBlock]
    max_turns: int = 0
    meta: dict[str, Any] | None = None


class PromptResponse(AcpModel):
    stop_reason: AcpStopReason
    meta: dict[str, Any] | None = None


# session/execute_shell


class ExecuteShellRequest(AcpModel):
    session_id: str
    command: str
    exclude_from_context: bool = False
    shell: str = "auto"
    meta: dict[str, Any] | None = None


class ExecuteShellResponse(AcpModel):
    exit_code: int
    cancelled: bool = False
    meta: dict[str, Any] | None = None


# session/execute_python


class ExecutePythonRequest(AcpModel):
    session_id: str
    code: str
    exclude_from_context: bool = False
    meta: dict[str, Any] | None = None


class ExecutePythonResponse(AcpModel):
    exit_code: int
    cancelled: bool = False
    meta: dict[str, Any] | None = None


# session/cancel_execution


class CancelExecutionRequest(AcpModel):
    session_id: str
    kind: str = "any"
    meta: dict[str, Any] | None = None


class CancelExecutionResponse(AcpModel):
    meta: dict[str, Any] | None = None


# session/cancel (notification)


class CancelNotification(AcpModel):
    session_id: str
    meta: dict[str, Any] | None = None


class CancelRequestNotification(AcpModel):
    request_id: str | int
    meta: dict[str, Any] | None = None


# session/set_mode


class SetSessionModeRequest(AcpModel):
    session_id: str
    mode_id: str
    meta: dict[str, Any] | None = None


class SetSessionModeResponse(AcpModel):
    meta: dict[str, Any] | None = None


# session/set_config_option


class SetSessionConfigOptionRequest(AcpModel):
    session_id: str
    config_id: str
    value: str
    meta: dict[str, Any] | None = None


class SetSessionConfigOptionResponse(AcpModel):
    config_options: list[dict]
    meta: dict[str, Any] | None = None


# session/rename


class RenameSessionRequest(AcpModel):
    session_id: str
    title: str
    meta: dict[str, Any] | None = None


class RenameSessionResponse(AcpModel):
    session: AcpSessionInfo
    meta: dict[str, Any] | None = None


# session/archive


class ArchiveSessionRequest(AcpModel):
    session_id: str
    archived: bool = True
    meta: dict[str, Any] | None = None


class ArchiveSessionResponse(AcpModel):
    session: AcpSessionInfo
    meta: dict[str, Any] | None = None


# session/delete


class DeleteSessionRequest(AcpModel):
    session_id: str
    force: bool = False
    meta: dict[str, Any] | None = None


class DeleteSessionResponse(AcpModel):
    deleted: bool
    meta: dict[str, Any] | None = None


# _mustang.agent/session/get_usage


class GetUsageRequest(AcpModel):
    session_id: str
    meta: dict[str, Any] | None = None


class TokenUsageSummary(AcpModel):
    input: int = 0
    output: int = 0
    cache_read: int = 0
    cache_write: int = 0
    total: int = 0


class ContextUsageSection(AcpModel):
    id: str
    label: str
    tokens: int
    percent: float


class ContextUsageSummary(AcpModel):
    total_tokens: int = 0
    context_window: int | None = None
    percent: float = 0.0
    sections: list[ContextUsageSection] = []


class HistoryUsageSummary(AcpModel):
    messages: int = 0
    turns: int = 0
    tool_calls: int = 0
    compactions: int = 0
    queued_turns: int = 0
    in_flight: bool = False
    last_run_at: str | None = None
    last_duration_ms: int | None = None


class MemoryUsageSummary(AcpModel):
    loaded: int = 0
    writable_scopes: int = 0


class EnvironmentUsageSummary(AcpModel):
    lsp_servers: list[str] = []
    mcp_servers: list[str] = []


class GetUsageResponse(AcpModel):
    session_id: str
    title: str | None = None
    cwd: str
    created_at: str | None = None
    updated_at: str | None = None
    model: str | None = None
    kernel_version: str
    tokens: TokenUsageSummary
    context: ContextUsageSummary
    history: HistoryUsageSummary
    memory: MemoryUsageSummary
    environment: EnvironmentUsageSummary
    cost_usd: float | None = None
    cost_note: str | None = None
    meta: dict[str, Any] | None = None
